Port zlib_available_p and zlib_decompress_region

harryfei opened the pull request:
Port decompress related functions.

There is already a test in `test/src/decompress-tests.el`. So I don't add a new one.

shaleh commented:
We have been moving the tests to the matching `rust_src` tree when we port.

shaleh commented on rust_src/src/lib.rs:
```diff
@@ -97,6 +98,7 @@ mod time;
 mod util;
 mod vectors;
 mod windows;
+mod decompress;
```
The list is current alphabetically ordered.

Wilfred commented on rust_src/src/decompress.rs:
```diff
@@ -0,0 +1,134 @@
+                return true;
+            }
+
+            // Decompress once successfully.
```
I don't entirely understand this comment -- could you rephrase it? I think at this point we've decoded the current data fine, but there's still some data left.

harryfei commented on src/emacs.c:
```diff
@@ -1507,10 +1507,6 @@ Using an Emacs configured with --with-x-toolkit=lucid does not have this problem
       syms_of_lcms2 ();
 #endif
 
-#ifdef HAVE_ZLIB
```
I think we can remove HAVE_ZLIB now. I will remove them in this PR update.

harryfei commented:
Fixed those:
* Move decompress tests to rust_src/src
* Remove all HAVE_ZLIB
* Change some comments
* Fixed the failed test
* Fixed mod declare order in `lib.rs`

@Wilfred @shaleh

shaleh commented on rust_src/src/decompress.rs:
```diff
@@ -0,0 +1,127 @@
+        // Zlib
+        0x78 => Box::new(ZlibDecoder::new(buffer)),
+        // Gzlib
+        0x1F => Box::new(GzDecoder::new(buffer)),
```
I think the comment reads better this way.

`// Assume the data is raw if neither zlib nor gzip header can be found.`

harryfei replied:
Fixed it. Thanks.

shaleh commented on rust_src/src/decompress.rs:
```diff
@@ -0,0 +1,127 @@
+            (iend - istart) as usize,
+        )
+    };
+
```
Spelling. The word is 'decompressor'.

harryfei replied:
Fixed it.

shaleh commented on rust_src/src/decompress.rs:
```diff
@@ -0,0 +1,127 @@
+                return true;
+            }
+
+            // Decompress one batch data successfully.
```
Typo. The word should be 'remaining'.

harryfei replied:
Fixed.

shaleh commented:
Looks good. I just left some grammar and spelling comments.
